parser-typechecker/src/Unison/Parser.hs

```diff
@@ -249,17 +210,16 @@ run' p s name env =
           then L.lexer name (trace (L.debugLex''' "lexer receives" s) s)
           else L.lexer name s
       pTraced = traceRemainingTokens "parser receives" *> p
-   in runParserT pTraced name (Input lex) env
+   in case runParserT pTraced name (Input lex) env of
+        Left err -> Left (Nel.head (P.bundleErrors err))
+        Right x -> Right x
 
 run :: Ord v => P v a -> String -> ParsingEnv -> Either (Err v) a
 run p s = run' p s ""
 
 -- Virtual pattern match on a lexeme.
 queryToken :: Ord v => (L.Lexeme -> Maybe a) -> P v (L.Token a)
-queryToken f = P.token go Nothing
-  where
-    go t@(f . L.payload -> Just s) = Right $ fmap (const s) t
-    go x = Left (pure (P.Tokens (x :| [])), Set.empty)
+queryToken f = P.token (traverse f) Set.empty
```

Comment:
Is this equivalent to the old version or did something change here?

At a glance it looks like the old version  returned an Either whereas this version seems to traverse with a Maybe, so I wonder if we're losing some error information or something?

It's not clear at a glance whether the traverse version is equivalent, but point-free expressions in ViewPatterns make my head hurt a bit so if you're confident this is correct I'll believe you 😋

Reply:
The type of `token` changed between [6.5.0](https://hackage.haskell.org/package/megaparsec-6.5.0/docs/Text-Megaparsec.html#v:token) and [7.0.0](https://hackage.haskell.org/package/megaparsec-7.0.0/docs/Text-Megaparsec.html#v:token). In 6.5.0 you provide a function `(Token s -> Either (Maybe (ErrorItem (Token s)), Set (ErrorItem (Token s))) a)` where the first element of the tuple returned by a Left are the unexpected tokens. 

`token` now accepts a function of type `(Token s -> Maybe a)` and if you return `Nothing` then the token is considered unexpected. So, we traverse the `Unison.Lexer.Token` to run `f` on the `Unison.Lexer.Lexeme`